### ISSUE TITLE
Liveness lints should consider statement attributes

### ISSUE THREAD

**Issue by mitaa** [rust-lang/rust#30326]
``` rust
#![feature(stmt_expr_attributes)]

fn main() {
    #[allow(unused_variables)]
    let mut foo = 7;
    #[allow(unused_assignments)]
    foo = 42;
}
```

```
<anon>:5:9: 5:16 warning: variable `foo` is assigned to, but never used, #[warn(unused_variables)] on by default
<anon>:5     let mut foo = 7;
                 ^~~~~~~
<anon>:7:5: 7:8 warning: value assigned to `foo` is never read, #[warn(unused_assignments)] on by default
<anon>:7     foo = 42;
             ^~~
```

I would expect to see no warnings.

**Comment by alexcrichton** [comment]
cc @Kimundi

**Comment by jonas-schievink** [comment]
The issue is that built-in lints are emitted by storing lint messages in a map keyed by some `NodeId`. The lint pass then traverses each item using an `IdVisitor`, which walks an item (ignoring nested items) and invokes a callback for all `NodeId`s. The callback then emits the stored lint messages if the current level allows that. Since no lint attributes are processed while the `IdVisitor` is doing its thing, the levels aren't changed by attributes applied to things within an item.

This issue has existed before, since lint attributes on match arms are ignored as well.

Fixing this either requires some tuning of the `IdVisitor` (maybe refactor how it works altogether) or doing the `NodeId` stuff manually. Or change how built-in lints are handled. Not sure what's the best way to do this.

**Comment by fhahn** [comment]
I also had a look at this issue and I thought it could be solved by extending `Session.add_lint` to also take and store the lint attribute of the node. Then when actually emitting the lints, all neccessary information should be available.

**Comment by jonas-schievink** [comment]
@fhahn You'd have to traverse parent blocks, since lint attributes could be defined on them, but then it could work. But at that point you'd be duplicating a lot of the logic of the lint contexts.

**Comment by Mark-Simulacrum** [comment]
This appears to have been fixed as of `rustc 1.18.0-nightly (bbdaad0dc 2017-04-14)`. Output from the code sample in the top comment is empty (no lint warnings).
